skipruntime-ts/server/src/server.ts

```diff
@@ -2,11 +2,80 @@ import { initService } from "skip-wasm";
 import type { SkipService, NamedCollections } from "skip-wasm";
 import { controlService, streamingService } from "./rest.js";
 
+/**
+ * Initialize and start a reactive Skip service.
+ *
+ * Calling `runService` will start a reactive service based on the `service` specification and `options`.
+ * The service offers two interfaces: a control API on `options.control_port`, and a streaming API on `options.streaming_port`.
+ *
+ * The service exposes resources specified by `service: SkipService`.
+ * Each resource has a name and identifies a collection that associates keys to values.
+ *
+ * The control API responds to the following HTTP requests:
+ *
+ * - `GET /v1/resources/:resource?param1=value1&...&paramN=valueN`:
+ *   Synchronous read of an entire resource.
+ *
+ *   Instantiate the named `resource` with parameters `{param1=value1,...,paramN=valueN}`, if necessary, and respond with the entire contents of the resource. The returned data will be a JSON-encoded value of type `[Json, Json[]][]`: an array of entries each of which associates a key to an array of its values.
+ *
+ * - `GET /v1/resources/:resource/:key?param1=value1&...&paramN=valueN`:
+ *   Synchronous read of a single key from a resource.
+ *
+ *   Instantiate the named `resource` with parameters `{param1=value1,...,paramN=valueN}`, if necessary, and respond with the values associated with the given `key`. The returned data will be a JSON-encoded value of type `Json[]`: an array of the `key`'s values.
+ *
+ * - `PATCH /v1/inputs/:collection`:
+ *   Partial write (update only the specified keys) of an input collection.
+ *
+ *   Update the named `collection` with the key-values entries in the request body.
+ *   The `collection` must be the name of one of the service's input collections, that is, one of the keys of the `Inputs` type parameter.
+ *   The body of the request will be parsed as JSON and interpreted as the `values` field of a `CollectionUpdate`, that is `[Json, Json[]][]`: an array of entries each of which associates a key to an array of its new values.
+ *
+ * - `PUT /v1/inputs/:collection/:key`:
+ *   Update of a single key of an input collection.
+ *
+ *   Update the named `collection` to associate the `key` with the values in the request body.
+ *   The `collection` must be the name of one of the service's input collections, that is, one of the keys of the `Inputs` type parameter.
+ *   The body of the request will be parsed as JSON and interpreted as `Json[]`: an array of the `key`'s new values.
+ *
+ * - `POST /v1/streams`:
+ *   Instantiate a resource and return a UUID to subscribe to updates.
+ *
+ *   Requires the request to have a `Content-Type: application/json` header.
+ *   The body of the request will be parsed as JSON and interpreted at type `{ resource: string, params: { [param: string]: string } }`.
+ *   Instantiate the named `resource` with parameters `params`, if necessary, and respond with a UUID that can be used to subscribe to updates.
```

Comment:
Something that isn't clear to me is why in this case we read the resource and params from the body, while in other cases they are in the URL. Is this a necessary inconsistency?

Reply:
Yes and no. GET requests cannot include a body, which leaves only the url (and headers) to pass data. This has the downside of making it difficult to distinguish a string from an int, and generally to pass non-primitive values.

Reply:
To expand on my previous reply, we could replace `GET /v1/resource/:resource` with a `POST /v1/snapshot` passing the `resource` and `params` (and possibly a set of `keys`) as a JSON object in the request body. This would also stress the fact that doing a sync read does indeed involve instantiating a resource.

Reply:
My feeling is that it would be better if all of the requests accepted the resource and params in the same way, both for consistency and because if something is hard/impossible in either of the two, then having both will leave users with the least common denominator. So if putting them in the body or the request is possible, then we would have more control over their parsing, etc., that seems like the better option to me.